pkg/arvo/neo/cod/std/src/fil/tasks.md

```diff
@@ -0,0 +1,624 @@
+      +:(,[%ud @ud] (rear pith))
+  =/  last-name=@ud  (rear sorted-names)
+  =/  name-missing=(list @ud)
+    %+  skim  (gulf 1 last-name)
+    |=  n=@ud
+    =(~ (find ~[n] sorted-names))
+  ?~  name-missing  +(last-name)
+  (rear name-missing)
+--
+::
+::  outer core
+^-  kook:neo
+|%
+::
+::  state is a %task
+++  state  pro/%task
+::
+::  we accept %gift and %task-diff pokes
+++  poke   (sy %task-diff %gift ~)
+::
+::  we define one generation of kids at
+::  /path/to/this/task/<@ud> which are
+::  also tasks like their parent, which
+::  means those kids also define one
+::  generation of kids below them, and so on
+++  kids
+  :+  ~  %y
+  %-  ~(gas by *lads:neo)
+  :~  :-  [|/%ud |]
+    [pro/%task (sy %task-diff %gift ~)]
+  ==
+++  deps
+  ::
+  ::  we don't need dependencies to negotiate
+  ::  state between parent and children tasks,
+  ::  but it would be valid to do so
+  *deps:neo
+++  form
+  ::
+  ::  inner core
+  ^-  form:neo
+  ::  XX refactor sample to same as other tutorials
+  |_  [=bowl:neo =aeon:neo stud:neo state-vase=vase]
+  ::  XX refactor this to default +init from counter
+  ++  init
+    |=  pal=(unit pail:neo)
+    ^-  (quip card:neo pail:neo)
+    `(need pal)
+  ++  poke
+    |=  [=stud:neo vax=vase]
+    ^-  (quip card:neo pail:neo)
+    =/  this  !<(task state-vase)
+    ?+    stud  !!
+        %gift
+      ?:  (check-kids bowl)
+        [~ task/!>(this(done %.y))]
+      [~ task/!>(this(done %.n))]
+    ::
+        %task-diff
+      =/  diff  !<(task-diff vax)
+      ?-    -.diff
+          %new
+        =/  name=@ud  (assign-name bowl)
+        =.  order.this
+          ?:  prepend.diff
+            ::  XX do we really need to do this? bad look if so
+            `(list pith)`[~[ud/name] order.this]
+          `(list pith)`(snoc order.this `pith`[ud/name ~])
+        =.  done.this  |
+        :_  task/!>(this)
+        :~  :-  (welp here.bowl ~[ud/name])
+            [%make %task `task/!>(task.diff) ~]
+        ==
+      ::
+          %edit
+        :-  ~
+        :-  %task
+        !>
+        %=  this
+          text  text.diff
+          done  ?:  (check-kids bowl)
+                  done.diff
+                %.n
+        ==
+      ::
+          %oust
+        =/  i  (find [pith.diff ~] order.this)
+        ?~  i  `task/!>(this)
+        :_  task/!>(this(order (oust [(need i) 1] order.this)))
+        :~  [(welp here.bowl pith.diff) [%tomb ~]]
+        ==
+      ::
+          %reorder
+        `task/!>(this(order order.diff))
+      ==
+    ==
+  --
+--
+```
+
+## The %gift poke
+In the `+poke` arm we declare this shrub takes a `%gift` as well as a `%task-diff`, but we don’t have to import `%gift` with a `/@` rune. This is a special poke like `%rely` that `/app/neo` gives us when another shrub’s state changes.
+
+Our shrub receives a `%gift` poke every time the state of one of its {kids / descendants} changes. Unlike `%rely`, we have to declare that we accept `%gift` pokes in the `+poke` arm.
```

Comment:
kids in the %.y case, descendants in the %.z case... I think?

Reply:
We **do** need to declare we accept %rely pokes in the poke arm as well

Reply:
`/imp/message-sub` takes a `%rely` poke and its `+poke` arm in `kook:neo` is `++  poke  ~`, which is what I'm referring to above. You're asserting the `%rely` poke in `form:neo`'s `+poke` arm.

However, `/imp/task` has to declare it takes a `%gift` poke otherwise tasks won't get marked as completed when all subtasks are complete. Whether or not it does declare `%gift`, on testing I do see complaints about a `%rely` poke when marking subtasks as complete.

```
no support for %rely
/app/neo/hoon:<[2.003 9].[2.003 50]>
/app/neo/hoon:<[1.997 9].[2.003 50]>
/app/neo/hoon:<[1.996 7].[2.009 20]>
/app/neo/hoon:<[1.995 7].[2.009 20]>
/app/neo/hoon:<[1.757 29].[1.757 48]>
/app/neo/hoon:<[1.757 21].[1.757 48]>
>>> [ %rely-nack
  [ship=~met pith=[i=%home t=[i=%tasks t=~[[%ud 2]]]]]
  ship=~met
    pith
  [ i=%srv
      t
    [ i=%hawk
        t
      ~[
        [ %uv
          0v1d2.6ptlj.u2uls.0opf2.k4bfm.vunfq.sntkq.eqb59.cu0ph.no71v.l096n.cg4q1.q0a29.top29.plplt.hgt85.83e6k.iff9p.e3c96.k6uca.m4ip3
        ]
      ]
    ]
  ]
]
```

Reply:
Sounds like we agree on the facts, I'd just phrase it differently. "Our shrub receives a `%gift` poke every time the state of one of its {kids / descendants} changes. **Like** in shrubs that take a `%rely`, we have to declare that we accept `%gift` pokes in the `+poke` arm.:

Reply:
Ah, I see you said exactly this in the latest update, nice.